### DISCUSSION TITLE
Live demo on Hugging Face

### DISCUSSION THREAD

**Discussion by radames** [lllyasviel/Fooocus#135] [Show and tell]
I made a live demo on Hugging Face using a Docker SDK
https://huggingface.co/spaces/SpacesExamples/Fooocus

Dockerfile if you want to run it locally
https://huggingface.co/spaces/SpacesExamples/Fooocus/blob/main/Dockerfile

**Comment by NashAiomos** [comment]
That's great, this is very helpful for me. My laptop can't run this program.

**Reply by radames** [reply]
@lllyasviel I'll be happy to link on the README.md if you think it's worth! thanks

**Reply by zyahav** [reply]
yes, please

**Reply by mashb1t** [reply]
@zyahav this space has been disabled, but you can use [Colab](https://github.com/lllyasviel/Fooocus?tab=readme-ov-file#colab) instead.

**Comment by oppenheimer-** [comment]
> 

I'm not sure if this works as intended. In basic mode as well as advanced mode, the results don't mirror the prompt in any way. No matter how detailed the prompt is. 

Look at "retro cyborg" for example.
![image (5)](https://github.com/lllyasviel/Fooocus/assets/10992675/5a52b70c-e631-4bae-bb91-abc71004d5d6)
![image (4)](https://github.com/lllyasviel/Fooocus/assets/10992675/00ef65bb-ea4e-45dd-8210-e71eafccdb61)
![image (3)](https://github.com/lllyasviel/Fooocus/assets/10992675/ed0c9a48-e5a8-4298-85f3-68326fe90305)
![image (2)](https://github.com/lllyasviel/Fooocus/assets/10992675/584407e1-1d07-4ceb-b48d-bc26038f0dd5)

**Reply by lllyasviel** [reply]
Sometimes huggingface space has a bug to give you images generated by other people...
Local test seem fine
![image](https://github.com/lllyasviel/Fooocus/assets/19834515/8a74a58b-15d1-4531-a98c-31703aa70c1e)

**Reply by radames** [reply]
yes indeed @lllyasviel , I notice that as well my guess is there is a global state leaking between different generations. I changed the concurrency queue to 1, but still something might be happening. It's more a Gradio concurrency issue, ideally you want to avoid global and use `gr.State` to keep individual session states

**Reply by oppenheimer-** [reply]
strange behaviour indeed. gotta try your docker.
thanks for taking a look at it! 🦾

**Reply by speaknowpotato** [reply]
hi @radames , do you have any updates on this issue? i also noticed this issue on my laptop, if i use `--listen` , people will get random pictures from others. thanks!

**Reply by radames** [reply]
hi @speaknowpotato, the issue is the Fooocus it's using a global state to run the pipeline workers, it probably needs a different state management to be ready for a shared web app, look at here https://github.com/lllyasviel/Fooocus/blob/09e0d1cb3ae5a1d74443009a41da9f96c1b54683/modules/async_worker.py#L4-L9

**Comment by zirlman** [comment]
Hey @radames . 

I've tried to clone your space and update the Dockerfile to the latest commit at the time of writing:
`cca0ca704a713ab153938e78de6787609c723cad` 

However, after building the project and running it. The space gets stuck at the building step. Are there any other steps required to run it properly? 

P.S. I've tried with `launch.py` but it also gets stuck 

<img width="1512" alt="image" src="https://github.com/lllyasviel/Fooocus/assets/24474083/ad5fb2f7-f73a-469f-9b07-27518cabe9ec">

Below you can see the updated Dockerfile:
```Dockerfile
FROM nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu22.04

ENV DEBIAN_FRONTEND=noninteractive \
    TZ=America/Los_Angeles

ARG USE_PERSISTENT_DATA

RUN apt-get update && apt-get install -y \
    git \
    make build-essential libssl-dev zlib1g-dev \
    libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm \
    libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev git-lfs  \
    ffmpeg libsm6 libxext6 cmake libgl1-mesa-glx \
    && rm -rf /var/lib/apt/lists/* \
    && git lfs install

WORKDIR /code

COPY ./requirements.txt /code/requirements.txt

# User
RUN useradd -m -u 1000 user
USER user
ENV HOME=/home/user \
    PATH=/home/user/.local/bin:$PATH

# Pyenv
RUN curl https://pyenv.run | bash
ENV PATH=$HOME/.pyenv/shims:$HOME/.pyenv/bin:$PATH

ARG PYTHON_VERSION=3.10.12
# Python
RUN pyenv install $PYTHON_VERSION && \
    pyenv global $PYTHON_VERSION && \
    pyenv rehash && \
    pip install --no-cache-dir --upgrade pip setuptools wheel && \
    pip install --no-cache-dir \
    datasets \
    huggingface-hub "protobuf<4" "click<8.1"

RUN pip install --no-cache-dir --upgrade -r /code/requirements.txt

# Set the working directory to /data if USE_PERSISTENT_DATA is set, otherwise set to $HOME/app
WORKDIR $HOME/app

# Copy the current directory contents into the container at $HOME/app setting the owner to the user

RUN git clone https://github.com/lllyasviel/Fooocus . && \
    # pin to a specific commit
    git checkout cca0ca704a713ab153938e78de6787609c723cad && \
    pip install --no-cache-dir xformers==0.0.20 triton==2.0.0 torch==2.0.1 torchvision==0.15.2 --extra-index-url https://download.pytorch.org/whl/cu118 && \
    pip install --no-cache-dir -r requirements_versions.txt

# remove the next line if you're running on your own GPU, it set max images to 3 and disables the API
RUN sed -i "s|image_number = gr.Slider(label='Image Number', minimum=1, maximum=32|image_number = gr.Slider(label='Image Number', minimum=1, maximum=3|" webui.py && \
    sed -i "s|shared.gradio_root = gr.Blocks(title='Fooocus ' + fooocus_version.version, css=modules.html.css).queue()|shared.gradio_root = gr.Blocks(title='Fooocus ' + fooocus_version.version, css=modules.html.css).queue(concurrency_count=1,api_open=False)|" webui.py && \
    sed -i "s|shared.gradio_root.launch(inbrowser=True, server_name=args.listen, server_port=args.port, share=args.share)|shared.gradio_root.launch(inbrowser=True, server_name=args.listen, server_port=args.port, share=args.share, show_api=False)|" webui.py

ENV HOME=/home/user \
    PATH=/home/user/.local/bin:$PATH \
    PYTHONPATH=$HOME/app \
    PYTHONUNBUFFERED=1 \
    GRADIO_ALLOW_FLAGGING=never \
    GRADIO_NUM_PORTS=1 \
    GRADIO_SERVER_NAME=0.0.0.0 \
    GRADIO_THEME=huggingface \
    SYSTEM=spaces

CMD ["python", "entry_with_update.py", "--preset", "realistic"]
```

**Reply by radames** [reply]
hi @zirlman, thanks for the heads up, I'll take a look, I've been a while since I've tested the Dockerfile with Fooocus, maybe they've changed something on their launching step that breaks the original Dockerfile

**Reply by ban-works** [reply]
It's really great that you made a space example, thank you. 
I have the same issue and I also noticed that the faceswap feature is gone. 
I went through the same process  as @zirlman and changed to the last commit, but somehow the ui does not contain the faceswap button.

My updated Dockerfile can be found below :

```
FROM nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu22.04

ENV DEBIAN_FRONTEND=noninteractive \
    TZ=America/Los_Angeles

ARG USE_PERSISTENT_DATA

RUN apt-get update && apt-get install -y \
    git \
    make build-essential libssl-dev zlib1g-dev \
    libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm \
    libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev git-lfs  \
    ffmpeg libsm6 libxext6 cmake libgl1-mesa-glx \
    && rm -rf /var/lib/apt/lists/* \
    && git lfs install

WORKDIR /code

COPY ./requirements.txt /code/requirements.txt

# User
RUN useradd -m -u 1000 user
USER user
ENV HOME=/home/user \
    PATH=/home/user/.local/bin:$PATH

# Pyenv
RUN curl https://pyenv.run | bash
ENV PATH=$HOME/.pyenv/shims:$HOME/.pyenv/bin:$PATH

ARG PYTHON_VERSION=3.10.12
# Python
RUN pyenv install $PYTHON_VERSION && \
    pyenv global $PYTHON_VERSION && \
    pyenv rehash && \
    pip install --no-cache-dir --upgrade pip setuptools wheel && \
    pip install --no-cache-dir \
    datasets \
    huggingface-hub "protobuf<4" "click<8.1"

RUN pip install --no-cache-dir --upgrade -r /code/requirements.txt

# Set the working directory to /data if USE_PERSISTENT_DATA is set, otherwise set to $HOME/app
WORKDIR $HOME/app

# Copy the current directory contents into the container at $HOME/app setting the owner to the user

RUN git clone https://github.com/lllyasviel/Fooocus . && \
    # pin to a specific commit
    git checkout cca0ca704a713ab153938e78de6787609c723cad && \
    pip install --no-cache-dir xformers==0.0.20 triton==2.0.0 torch==2.0.1 torchvision==0.15.2 --extra-index-url https://download.pytorch.org/whl/cu118 && \
    pip install --no-cache-dir -r requirements_versions.txt

# remove the next line if you're running on your own GPU, it set max images to 3 and disables the API
RUN sed -i "s|image_number = gr.Slider(label='Image Number', minimum=1, maximum=32|image_number = gr.Slider(label='Image Number', minimum=1, maximum=3|" webui.py && \
    sed -i "s|shared.gradio_root = gr.Blocks(title='Fooocus ' + fooocus_version.version, css=modules.html.css).queue()|shared.gradio_root = gr.Blocks(title='Fooocus ' + fooocus_version.version, css=modules.html.css).queue(concurrency_count=1,api_open=False)|" webui.py && \
    sed -i "s|shared.gradio_root.launch(inbrowser=True, server_name=args.listen, server_port=args.port, share=args.share)|shared.gradio_root.launch(inbrowser=True, server_name=args.listen, server_port=args.port, share=args.share, show_api=False)|" webui.py

ENV HOME=/home/user \
    PATH=/home/user/.local/bin:$PATH \
    PYTHONPATH=$HOME/app \
    PYTHONUNBUFFERED=1 \
    GRADIO_ALLOW_FLAGGING=never \
    GRADIO_NUM_PORTS=1 \
    GRADIO_SERVER_NAME=0.0.0.0 \
    GRADIO_THEME=huggingface \
    SYSTEM=spaces

CMD ["python", "launch.py"]
```

**Reply by zirlman** [reply]
@ban-works I eventually managed to setup the app by using a different commit hash.  However, as you noted I'm also missing the FaceSwap feature. Maybe the `launch.py` doesn't support face swap out of the box?

**Reply by zirlman** [reply]
Another thing that I found during trying to setup is that there are some missing styles. 
Local environment:
<img width="469" alt="Screenshot 2023-12-02 at 14 34 35" src="https://github.com/lllyasviel/Fooocus/assets/24474083/400e75bf-a536-4720-a9f3-7f41f012b1bb">
vs Fooocus  space (same as on my space)

<img width="462" alt="image" src="https://github.com/lllyasviel/Fooocus/assets/24474083/2d787e2d-958e-4312-9c3f-e402d0fa58dc">

Maybe there some configurations that are disabled in this space 🤔

**Reply by ShreyamMaity** [reply]
![image](https://github.com/lllyasviel/Fooocus/assets/38105595/8edd16fe-f287-4600-ad9d-cc29b5afb8d2)
That's because the Dockerfile is pointing to a specific commit only, And that commit might not have the newest settings.

**Reply by zirlman** [reply]
> ![image](https://github.com/lllyasviel/Fooocus/assets/38105595/8edd16fe-f287-4600-ad9d-cc29b5afb8d2)
> That's because the Dockerfile is pointing to a specific commit only, And that commit might not have the newest settings.

Thank you for pointing that out. It's true, @radames 's commit in the repo doesn't include the FaceSwap feature. However, removing the checkout line and building the latest version of Fooocus doesn't provide all the features on HF as it would locally. 

Maybe I'm missing something, did you try to create a HF space without that line? If so, did you succeed?

**Reply by radames** [reply]
yes, I've pinned the version in case @main gets an error and sometimes the Space boost up. Anyway, I've removed the pinned version, and now it's on main.

**Reply by Maniaz09** [reply]
@zirlman how did you manage to setup the app?
i could use some help.

**Reply by zirlman** [reply]
I ended up with the following Dockerfile: 

```docker
FROM nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu22.04

ENV DEBIAN_FRONTEND=noninteractive \
    TZ=America/Los_Angeles

ARG USE_PERSISTENT_DATA

RUN apt-get update && apt-get install -y \
    git \
    make build-essential libssl-dev zlib1g-dev \
    libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm \
    libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev git-lfs  \
    ffmpeg libsm6 libxext6 cmake libgl1-mesa-glx \
    && rm -rf /var/lib/apt/lists/* \
    && git lfs install

WORKDIR /code

COPY ./requirements.txt /code/requirements.txt

# User
RUN useradd -m -u 1000 user
USER user
ENV HOME=/home/user \
    PATH=/home/user/.local/bin:$PATH

# Pyenv
RUN curl https://pyenv.run | bash
ENV PATH=$HOME/.pyenv/shims:$HOME/.pyenv/bin:$PATH

ARG PYTHON_VERSION=3.10.12
# Python
RUN pyenv install $PYTHON_VERSION && \
    pyenv global $PYTHON_VERSION && \
    pyenv rehash && \
    pip install --no-cache-dir --upgrade pip setuptools wheel && \
    pip install --no-cache-dir \
    datasets \
    huggingface-hub "protobuf<4" "click<8.1"

RUN pip install --no-cache-dir --upgrade -r /code/requirements.txt

# Set the working directory to /data if USE_PERSISTENT_DATA is set, otherwise set to $HOME/app
WORKDIR $HOME/app

# Copy the current directory contents into the container at $HOME/app setting the owner to the user

RUN git clone https://github.com/lllyasviel/Fooocus . && \
    # pin to a specific commit
    git checkout bd4d40203c7e727fd06fc58ec122c25befcc5644 && \
    pip install --no-cache-dir xformers==0.0.20 triton==2.0.0 torch==2.0.1 torchvision==0.15.2 --extra-index-url https://download.pytorch.org/whl/cu118 && \
    pip install --no-cache-dir -r requirements_versions.txt

# remove the next line if you're running on your own GPU, it set max images to 3 and disables the API
RUN sed -i "s|image_number = gr.Slider(label='Image Number', minimum=1, maximum=32|image_number = gr.Slider(label='Image Number', minimum=1, maximum=3|" webui.py && \
    sed -i "s|shared.gradio_root = gr.Blocks(title='Fooocus ' + fooocus_version.version, css=modules.html.css).queue()|shared.gradio_root = gr.Blocks(title='Fooocus ' + fooocus_version.version, css=modules.html.css).queue(concurrency_count=1,api_open=False)|" webui.py && \
    sed -i "s|shared.gradio_root.launch(inbrowser=True, server_name=args.listen, server_port=args.port, share=args.share)|shared.gradio_root.launch(inbrowser=True, server_name=args.listen, server_port=args.port, share=args.share, show_api=False)|" webui.py

ENV HOME=/home/user \
    PATH=/home/user/.local/bin:$PATH \
    PYTHONPATH=$HOME/app \
    PYTHONUNBUFFERED=1 \
    GRADIO_ALLOW_FLAGGING=never \
    GRADIO_NUM_PORTS=1 \
    GRADIO_SERVER_NAME=0.0.0.0 \
    GRADIO_THEME=huggingface \
    SYSTEM=spaces

CMD ["python", "entry_with_update.py", "--listen"]
```
